internal/pwru/skb_metadata.go

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: Apache-2.0
+/* Copyright Leon Hwang */
+/* Copyright Authors of Cilium */
+
+package pwru
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/Asphaltt/mybtf"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/asm"
+	"github.com/cilium/ebpf/btf"
+	"github.com/leonhwangprojects/bice"
+)
+
+const (
+	labelSetSkbMetadataExit = "__set_skb_metadata_exit"
+	setSkbMetadataStub      = "set_skb_metadata"
+	maxSkbMetadata          = 4
+)
+
+type SkbMetadata struct {
+	expr string
+	last string
+	t    btf.Type // type of last field
+	insn asm.Instructions
+}
+
+func isChar(c byte) bool {
+	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
+}
+
+func isDigit(c byte) bool {
+	return c >= '0' && c <= '9'
+}
+
+func isValidChar(c byte) bool {
+	return isChar(c) || c == '_' || isDigit(c)
+}
+
+func newSkbMetadata(expr string, spec *btf.Spec) (*SkbMetadata, error) {
+	var md SkbMetadata
+	md.expr = expr
+
+	if !strings.HasPrefix(expr, "skb") {
+		return nil, errors.New("--output-skb-metadata must starts with skb")
+	}
+
+	for i := len(expr) - 1; i >= 0; i-- {
+		if !isValidChar(expr[i]) {
```

Comment:
Why can't we return an error instead?

Reply:
This `for` is to get the last attribute name, e.g. 'ifindex' for 'skb->dev->ifindex'.
Afterwards, while outputting its value, the output should be 'ifindex=XX' instead of 'skb->dev->ifindex=XX'.

Besides it, I should add a `for` to validate all chars.